# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [2]:
os.getcwd()
os.chdir('../data/')
os.getcwd()
df = pd.read_csv('step3_output.csv')
print(df.head())

                            Name       state  summit_elev  vertical_drop  \
0               Hilltop Ski Area      Alaska         2090            294   
1            Sunrise Park Resort     Arizona        11100           1800   
2  Yosemite Ski & Snowboard Area  California         7800            600   
3               Donner Ski Ranch  California         8012            750   
4                  June Mountain  California        10090           2590   

   trams  fastEight  fastSixes  fastQuads  quad  triple  ...  \
0      0        0.0          0          0     0       1  ...   
1      0        0.0          0          1     2       3  ...   
2      0        0.0          0          0     0       1  ...   
3      0        0.0          0          0     0       1  ...   
4      0        0.0          0          2     0       0  ...   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0               30.0            30.0        150.000000       36.0   
1              800.0

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [3]:
df_state=df[['state']]
df = pd.concat([df.drop(df_state, axis=1), pd.get_dummies(df_state)], axis=1)
print(df.head())

                            Name  summit_elev  vertical_drop  trams  \
0               Hilltop Ski Area         2090            294      0   
1            Sunrise Park Resort        11100           1800      0   
2  Yosemite Ski & Snowboard Area         7800            600      0   
3               Donner Ski Ranch         8012            750      0   
4                  June Mountain        10090           2590      0   

   fastEight  fastSixes  fastQuads  quad  triple  double  ...  \
0        0.0          0          0     0       1       0  ...   
1        0.0          0          1     2       3       1  ...   
2        0.0          0          0     0       1       3  ...   
3        0.0          0          0     0       1       5  ...   
4        0.0          0          2     0       0       4  ...   

   state_Rhode Island  state_South Dakota  state_Tennessee  state_Utah  \
0                   0                   0                0           0   
1                   0             

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [4]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df['AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [5]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [6]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [7]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [8]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model

In [12]:
evs = explained_variance_score(y_test,y_pred)
mse = mean_absolute_error(y_test,y_pred)

print('Model 1: Explained Variance Score is ' + str(evs))

print('Model 1: Mean Absolute Error is ' + str(mse))

Model 1: Explained Variance Score is -9.157466294140797e+22
Model 1: Mean Absolute Error is 1097941511684.7526


**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [13]:
print('Model 1: Intercept is ' + str(lm.intercept_))

Model 1: Intercept is 276129002993.6294


**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [14]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values(by=['Coefficient'],ascending=False)

Coefficient
trams                 3.347072e+13
fastEight             2.969728e+13
fastSixes             2.071183e+13
total_chairs          2.027535e+13
surface               1.158275e+13
double                1.120901e+13
triple                9.565921e+12
quad                  5.784328e+12
state_New York        4.880748e+12
fastQuads             4.599856e+12
state_Michigan        4.394871e+12
state_New Hampshire   3.812742e+12
state_Wisconsin       3.646940e+12
state_Pennsylvania    3.646940e+12
state_California      3.282633e+12
state_Minnesota       3.282633e+12
state_Vermont         3.080263e+12
state_New Mexico      2.860667e+12
state_Idaho           2.860667e+12
state_Wyoming         2.619517e+12
state_Colorado        2.619517e+12
state_Maine           2.619517e+12
state_Connecticut     2.619517e+12
state_Ohio            2.350188e+12
state_Washington      2.350188e+12
state_Massachusetts   2.350188e+12
state_Montana         2.041556e+12
state_Utah            2.041556e+12
state_Oregon          2.041556e+12
state_Iowa            2.041556e+12
state_Illinois        2.041556e+12
state_West Virginia   2.041556e+12
state_North Carolina  2.041556e+12
state_Indiana         1.671998e+12
state_Missouri        1.671998e+12
state_Virginia        1.671998e+12
state_South Dakota    1.185858e+12
state_Arizona         1.185858e+12
state_Alaska          1.185858e+12
state_Nevada          1.185858e+12
state_Tennessee       1.185858e+12
state_Maryland        9.862935e+07
state_New Jersey      3.636598e+04
AdultWeekday          9.756836e+00
summit_elev           2.254457e+00
Snow Making_ac        1.808594e+00
Runs                  1.173340e+00
clusters              9.921875e-01
vertical_drop         8.893484e-01
LongestRun_mi         8.642578e-01
NightSkiing_ac        7.539062e-01
TerrainParks          6.054688e-01
yearsOpen             5.683594e-01
state_Rhode Island    4.376831e-01
SkiableTerrain_ac     4.160156e-01
daysOpenLastYear      3.613281e-01
projectedDaysOpen     3.105469e-01
averageSnowfall       2.500000e-01

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [15]:
df = pd.read_csv('step3_output.csv')
df = df.drop('state', axis=1)
print(df.head())

                            Name  summit_elev  vertical_drop  trams  \
0               Hilltop Ski Area         2090            294      0   
1            Sunrise Park Resort        11100           1800      0   
2  Yosemite Ski & Snowboard Area         7800            600      0   
3               Donner Ski Ranch         8012            750      0   
4                  June Mountain        10090           2590      0   

   fastEight  fastSixes  fastQuads  quad  triple  double  ...  \
0        0.0          0          0     0       1       0  ...   
1        0.0          0          1     2       3       1  ...   
2        0.0          0          0     0       1       3  ...   
3        0.0          0          0     0       1       5  ...   
4        0.0          0          2     0       0       4  ...   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0               30.0            30.0        150.000000       36.0   
1              800.0            80.0        

In [16]:
# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df['AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

y_pred = model.predict(X_test)

evs = explained_variance_score(y_test,y_pred)
mse = mean_absolute_error(y_test,y_pred)

print('Model 2: Explained Variance Score is ' + str(evs))

print('Model 2: Mean Absolute Error is ' + str(mse))

print('Model 2: Intercept is ' + str(lm.intercept_))

pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values(by=['Coefficient'],ascending=False)

Model 2: Explained Variance Score is 0.5318037556069808
Model 2: Mean Absolute Error is 7.104478104298501
Model 2: Intercept is 56.21744716908226


Coefficient
AdultWeekday       1.060707e+01
clusters           3.037833e+00
summit_elev        2.983387e+00
daysOpenLastYear   1.804732e+00
averageSnowfall    1.558668e+00
Snow Making_ac     1.535561e+00
vertical_drop      1.470070e+00
triple             1.262915e+00
surface            1.045684e+00
Runs               9.106585e-01
projectedDaysOpen  4.977379e-01
double             4.417471e-01
LongestRun_mi      4.347729e-01
NightSkiing_ac     3.978057e-01
total_chairs       3.381461e-01
quad               3.154156e-01
yearsOpen          1.041218e-01
TerrainParks       7.346146e-02
fastQuads          2.412343e-02
SkiableTerrain_ac  2.123826e-02
trams              8.881784e-16
fastSixes          2.220446e-16
fastEight          0.000000e+00

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [17]:
df = pd.read_csv('step3_output.csv')
df = df.drop(['state','summit_elev'], axis=1)
print(df.head())

                            Name  vertical_drop  trams  fastEight  fastSixes  \
0               Hilltop Ski Area            294      0        0.0          0   
1            Sunrise Park Resort           1800      0        0.0          0   
2  Yosemite Ski & Snowboard Area            600      0        0.0          0   
3               Donner Ski Ranch            750      0        0.0          0   
4                  June Mountain           2590      0        0.0          0   

   fastQuads  quad  triple  double  surface  ...  SkiableTerrain_ac  \
0          0     0       1       0        2  ...               30.0   
1          1     2       3       1        0  ...              800.0   
2          0     0       1       3        1  ...               88.0   
3          0     0       1       5        2  ...              505.0   
4          2     0       0       4        1  ...             1500.0   

   Snow Making_ac  daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  \
0          

In [18]:
# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df['AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

y_pred = model.predict(X_test)

evs = explained_variance_score(y_test,y_pred)
mse = mean_absolute_error(y_test,y_pred)

print('Model 3: Explained Variance Score is ' + str(evs))

print('Model 3: Mean Absolute Error is ' + str(mse))

print('Model 3: Intercept is ' + str(lm.intercept_))

pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values(by=['Coefficient'],ascending=False)

Model 3: Explained Variance Score is 0.4977728038168997
Model 3: Mean Absolute Error is 7.601271928640627
Model 3: Intercept is 56.161785482548176


Coefficient
AdultWeekday       1.020135e+01
clusters           2.533290e+00
Snow Making_ac     2.343665e+00
averageSnowfall    2.209559e+00
vertical_drop      2.003330e+00
daysOpenLastYear   1.278066e+00
SkiableTerrain_ac  1.270678e+00
triple             1.116618e+00
surface            7.377654e-01
double             5.870431e-01
total_chairs       4.574802e-01
LongestRun_mi      3.896662e-01
NightSkiing_ac     3.834888e-01
TerrainParks       3.572860e-01
quad               2.008690e-01
projectedDaysOpen  1.382609e-01
fastQuads          1.310097e-01
Runs               9.613394e-02
yearsOpen          5.402929e-02
fastSixes          1.332268e-15
trams              8.881784e-16
fastEight          4.440892e-16

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | -9.157466294140797e+22 | 1097941511684.7526  |-|
| Model 2. | 0.5318037556069808 | |'state'|
| Model 3. | 0.4977728038168997 | 7.601271928640627 |'state','summit_elev','base_elev'|

Model Selection: